File: parsers/python/normalize.py
#!/usr/bin/env python3
"""
Normalizes Python source files to the language-agnostic NStmt JSON format.

Usage:
    python normalize.py <source-root>

Output (stdout):
    {
      "functions": [
        { "file": "rel/path.py", "name": "fn_name", "line": 10, "body_ast": { ... } },
        ...
      ]
    }

Mapping (ast node → NStmt kind):
    ast.If          → "if"
    ast.For         → "for"
    ast.While       → "while"
    ast.Call        → "call"    (when the whole expression is a call)
    ast.Return      → "return"
    ast.Try         → "try"
    ast.Raise       → "throw"
    ast.Break       → "break"
    ast.Continue    → "continue"
    ast.FunctionDef → "block"   (body is normalized recursively)
    everything else → "expr"
"""

import ast
import json
import sys
from pathlib import Path


# ---------------------------------------------------------------------------
# Expression helpers
# ---------------------------------------------------------------------------

def _unparse(node) -> str:
    """Return a compact source representation of an AST expression/statement."""
    try:
        return ast.unparse(node)
    except Exception:
        return type(node).__name__


# ---------------------------------------------------------------------------
# Statement normalization
# ---------------------------------------------------------------------------

def norm_stmts(stmts: list) -> dict:
    """Wrap a list of statements in an NStmt::Block node."""
    return {"kind": "block", "body": [norm_stmt(s) for s in stmts]}


def norm_stmt(node) -> dict:
    """Convert a single Python AST statement node to an NStmt dict."""

    # ── if ────────────────────────────────────────────────────────────────────
    if isinstance(node, ast.If):
        else_branch = norm_stmts(node.orelse) if node.orelse else None
        return {
            "kind":        "if",
            "condition":   _unparse(node.test),
            "then_branch": norm_stmts(node.body),
            "else_branch": else_branch,
        }

    # ── for ───────────────────────────────────────────────────────────────────
    if isinstance(node, ast.For):
        return {
            "kind":      "for",
            "init":      f"{_unparse(node.target)} in {_unparse(node.iter)}",
            "condition": None,
            "update":    None,
            "body":      norm_stmts(node.body),
        }

    # ── while ─────────────────────────────────────────────────────────────────
    if isinstance(node, ast.While):
        return {
            "kind":      "while",
            "condition": _unparse(node.test),
            "body":      norm_stmts(node.body),
        }

    # ── return ────────────────────────────────────────────────────────────────
    if isinstance(node, ast.Return):
        return {
            "kind":  "return",
            "value": _unparse(node.value) if node.value else None,
        }

    # ── call (bare expression whose value is a Call) ──────────────────────────
    if isinstance(node, ast.Expr) and isinstance(node.value, ast.Call):
        return {
            "kind":   "call",
            "target": _unparse(node.value.func),
        }

    # ── try (Python 3.11+ splits Try into TryStar; handle both) ──────────────
    if isinstance(node, (ast.Try,)) or (
        hasattr(ast, "TryStar") and isinstance(node, ast.TryStar)
    ):
        catch_stmts = []
        for handler in node.handlers:
            catch_stmts.extend(handler.body)

        finally_body = getattr(node, "finalbody", None)
        return {
            "kind":          "try",
            "try_block":     norm_stmts(node.body),
            "catch_block":   norm_stmts(catch_stmts) if catch_stmts else None,
            "finally_block": norm_stmts(finally_body) if finally_body else None,
        }

    # ── raise ─────────────────────────────────────────────────────────────────
    if isinstance(node, ast.Raise):
        return {
            "kind":  "throw",
            "value": _unparse(node.exc) if node.exc else None,
        }

    # ── break / continue ──────────────────────────────────────────────────────
    if isinstance(node, ast.Break):
        return {"kind": "break"}

    if isinstance(node, ast.Continue):
        return {"kind": "continue"}

    # ── nested function / async function — recurse into body ──────────────────
    if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
        return norm_stmts(node.body)

    # ── fallback ──────────────────────────────────────────────────────────────
    return {"kind": "expr", "text": _unparse(node)}


# ---------------------------------------------------------------------------
# File-level collection
# ---------------------------------------------------------------------------

def collect_functions(tree: ast.AST, rel_path: str) -> list:
    """Return one entry per function/method found anywhere in the module."""
    results = []
    for node in ast.walk(tree):
        if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
            results.append({
                "file":     rel_path,
                "name":     node.name,
                "line":     node.lineno,
                "body_ast": norm_stmts(node.body),
            })
    return results


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

def main() -> None:
    if len(sys.argv) < 2:
        print("Usage: normalize.py <source-root>", file=sys.stderr)
        sys.exit(1)

    root = Path(sys.argv[1]).resolve()
    all_functions: list = []

    for path in sorted(root.rglob("*.py")):
        try:
            source = path.read_text(encoding="utf-8")
            tree   = ast.parse(source, filename=str(path))
            rel    = path.relative_to(root).as_posix()
            all_functions.extend(collect_functions(tree, rel))
        except Exception as exc:
            print(f"[warn] {path}: {exc}", file=sys.stderr)

    print(json.dumps({"functions": all_functions}, indent=2))


if __name__ == "__main__":
    main()

File: parsers/go/normalize.go
// go-normalize — converts Go source files to the language-agnostic NStmt JSON.
//
// Usage:
//
//	go-normalize <source-root>
//
// Output (stdout):
//
//	{ "functions": [ { "file": "...", "name": "...", "line": N, "body_ast": {...} }, ... ] }
//
// Mapping (go/ast node → NStmt kind):
//
//	*ast.IfStmt      → "if"
//	*ast.ForStmt     → "for"    (classic for; also covers while-style loops)
//	*ast.RangeStmt   → "for"
//	*ast.ReturnStmt  → "return"
//	*ast.ExprStmt    → "call"   (when the expression is a *ast.CallExpr)
//	*ast.BlockStmt   → "block"
//	BranchStmt break → "break"
//	BranchStmt cont  → "continue"
//	everything else  → "expr"

package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"go/ast"
	"go/parser"
	"go/printer"
	"go/token"
	"os"
	"path/filepath"
	"strings"
)

// ---------------------------------------------------------------------------
// NStmt representation
// ---------------------------------------------------------------------------

// NStmt is a generic map that serializes to the tagged-object format expected
// by the Rust NStmt enum: { "kind": "if", ... }
type NStmt map[string]interface{}

// ---------------------------------------------------------------------------
// Expression / statement helpers
// ---------------------------------------------------------------------------

func exprStr(e ast.Expr, fset *token.FileSet) string {
	if e == nil {
		return ""
	}
	var buf bytes.Buffer
	if err := printer.Fprint(&buf, fset, e); err != nil {
		return fmt.Sprintf("%T", e)
	}
	return buf.String()
}

func stmtStr(s ast.Stmt, fset *token.FileSet) string {
	if s == nil {
		return ""
	}
	var buf bytes.Buffer
	if err := printer.Fprint(&buf, fset, s); err != nil {
		return fmt.Sprintf("%T", s)
	}
	return strings.TrimSpace(buf.String())
}

// ---------------------------------------------------------------------------
// Normalization
// ---------------------------------------------------------------------------

func normBlock(stmts []ast.Stmt, fset *token.FileSet) NStmt {
	body := make([]NStmt, 0, len(stmts))
	for _, s := range stmts {
		if n := normStmt(s, fset); n != nil {
			body = append(body, n)
		}
	}
	return NStmt{"kind": "block", "body": body}
}

func wrapBlock(s ast.Stmt, fset *token.FileSet) NStmt {
	if bs, ok := s.(*ast.BlockStmt); ok {
		return normBlock(bs.List, fset)
	}
	n := normStmt(s, fset)
	body := []NStmt{}
	if n != nil {
		body = append(body, n)
	}
	return NStmt{"kind": "block", "body": body}
}

func normStmt(s ast.Stmt, fset *token.FileSet) NStmt {
	if s == nil {
		return nil
	}

	switch n := s.(type) {

	// ── if ────────────────────────────────────────────────────────────────────
	case *ast.IfStmt:
		var elseB interface{}
		if n.Else != nil {
			elseB = normStmt(n.Else, fset)
		}
		return NStmt{
			"kind":        "if",
			"condition":   exprStr(n.Cond, fset),
			"then_branch": normBlock(n.Body.List, fset),
			"else_branch": elseB,
		}

	// ── for (classic: init; cond; post) ──────────────────────────────────────
	case *ast.ForStmt:
		var init, cond, update interface{}
		if n.Init != nil {
			init = stmtStr(n.Init, fset)
		}
		if n.Cond != nil {
			cond = exprStr(n.Cond, fset)
		}
		if n.Post != nil {
			update = stmtStr(n.Post, fset)
		}
		return NStmt{
			"kind":      "for",
			"init":      init,
			"condition": cond,
			"update":    update,
			"body":      normBlock(n.Body.List, fset),
		}

	// ── for … range ───────────────────────────────────────────────────────────
	case *ast.RangeStmt:
		init := fmt.Sprintf("%s := range %s",
			exprStr(n.Key, fset), exprStr(n.X, fset))
		return NStmt{
			"kind":      "for",
			"init":      init,
			"condition": nil,
			"update":    nil,
			"body":      normBlock(n.Body.List, fset),
		}

	// ── return ────────────────────────────────────────────────────────────────
	case *ast.ReturnStmt:
		var val interface{}
		if len(n.Results) > 0 {
			val = exprStr(n.Results[0], fset)
		}
		return NStmt{"kind": "return", "value": val}

	// ── call (expression statement whose expression is a call) ────────────────
	case *ast.ExprStmt:
		if call, ok := n.X.(*ast.CallExpr); ok {
			return NStmt{"kind": "call", "target": exprStr(call.Fun, fset)}
		}
		return NStmt{"kind": "expr", "text": exprStr(n.X, fset)}

	// ── block ─────────────────────────────────────────────────────────────────
	case *ast.BlockStmt:
		return normBlock(n.List, fset)

	// ── break / continue ──────────────────────────────────────────────────────
	case *ast.BranchStmt:
		switch n.Tok {
		case token.BREAK:
			return NStmt{"kind": "break"}
		case token.CONTINUE:
			return NStmt{"kind": "continue"}
		}

		// ── if-init block (e.g. `if x := f(); x != nil { ... }`) ─────────────────
		// handled above via *ast.IfStmt; the init is included in condition text.
	}

	// Fallback
	return NStmt{"kind": "expr", "text": stmtStr(s, fset)}
}

// ---------------------------------------------------------------------------
// Per-function entry
// ---------------------------------------------------------------------------

type FuncEntry struct {
	File    string      `json:"file"`
	Name    string      `json:"name"`
	Line    int         `json:"line"`
	BodyAst interface{} `json:"body_ast"`
}

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "Usage: go-normalize <source-root>")
		os.Exit(1)
	}

	root := os.Args[1]
	var functions []FuncEntry
	fset := token.NewFileSet()

	err := filepath.Walk(root, func(path string, info os.FileInfo, walkErr error) error {
		if walkErr != nil || info.IsDir() {
			return walkErr
		}
		if !strings.HasSuffix(path, ".go") {
			return nil
		}

		f, parseErr := parser.ParseFile(fset, path, nil, 0)
		if parseErr != nil {
			fmt.Fprintf(os.Stderr, "[warn] %s: %v\n", path, parseErr)
			return nil
		}

		rel, relErr := filepath.Rel(root, path)
		if relErr != nil {
			rel = path
		}
		rel = filepath.ToSlash(rel)

		for _, decl := range f.Decls {
			fd, ok := decl.(*ast.FuncDecl)
			if !ok || fd.Body == nil {
				continue
			}
			pos := fset.Position(fd.Pos())
			functions = append(functions, FuncEntry{
				File:    rel,
				Name:    fd.Name.Name,
				Line:    pos.Line,
				BodyAst: normBlock(fd.Body.List, fset),
			})
		}
		return nil
	})

	if err != nil {
		fmt.Fprintf(os.Stderr, "[error] walk: %v\n", err)
		os.Exit(1)
	}

	if functions == nil {
		functions = []FuncEntry{}
	}

	out, _ := json.MarshalIndent(map[string]interface{}{"functions": functions}, "", "  ")
	os.Stdout.Write(out)
	fmt.Println()
}
